### ISSUE TITLE
Strange right arrow completion

### ISSUE THREAD

**Issue by oscarbenjamin** [ipython/ipython#14137]
I don't really know how to explain a full reproducer for this because it depends on the history of things that I have typed into ipython. Right now I can reproduce this as follows but I am not sure what previous things I have done that cause this to happen.

First just setup a few pieces:
```python
In [4]: solve = lambda *a, **kw: None

In [5]: syms = E = None
```
Now I run this command:
```python
In [6]: solve(E, syms)
```
Now I want to rerun this but change it to `solve(groebner(E), syms)` so I follow these steps carefully:

Push `<up>`, then `<home>` then `<right>` 6 times and type `groebner(`. Now I have this:
```python
In [7]: solve(groebner(E, syms)
```
and my cursor is flashing on the `E`.

I want to add the closing bracket `)` after the `E` so I press `<right>` and then something weird happens:
```python
In [7]: solve(groebner(, syms, dict=True)E, syms)
```
Somehow pressing `<right>` inserted the text `, syms, dict=True)` and then left my cursor at the end of that text.

Is this a bug? Have I perhaps somehow accidentally configured something that makes this happen?

Adding the text `, syms, dict=True)` is certainly something that I have previously done as an edit so maybe `ipython` "remembers" that somehow?

Note that if I change anything about this like if I delete the `s` from the start of the line then it does not happen. Somehow this is specific to the exact text that I am editing e.g. if I have this then I can use both `<left>` and `<right>` normally:
```
In [7]: olve(groebner(E, syms)
```
I think it is trying to complete from the history somehow e.g. if I have this:
```
In [7]: solve(groebner(E, syms),
```
then the completer offers:
![image](https://github.com/ipython/ipython/assets/1159732/19f4441f-a798-4d13-9145-e7214af4d81f)
It seems that the completion is being used when the cursor is not at the end of the line somehow.

**Comment by codtiger** [comment]
Running into the same issue on WSL2 / Windows 10. Pretty strange behavior that right arrow at some random points triggers auto-completion mid line, while the expected behavior is to only complete when the cursor's on the end character and a suggestion is shown. Has anyone been trying to look at this or try to replicate it?
Python Version: 3.10.12
IPython Version: 8.7.12

**Comment by iftekhart** [comment]
Reporting the same issue in ipython version 8.22.2

**Comment by k4lizen** [comment]
repro:
```python
In [15]: 123
Out[15]: 123

In [16]: 132
```
Enter `123` in, press Enter. Then write down `2`. Press Left. Paste/insert `1` from clipboard (in front of the `2`). Press right arrow key. A `3` is autocompleted and it results in 132.

**Comment by oscarbenjamin** [comment]
> repro:

Thanks, I can reproduce it with that as well. I guess it sees that `12` can be completed to `123` and so inserts `3` but the cursor is in the wrong position.

It only happens if the cursor is not at the end of the line when you add the `1` so that the match occurs while the cursor is in the wrong position.

**Comment by t4rf9** [comment]
Same problem!
I wish it would be possible to disable auto completion while the cursor is not at the end, if there is no other particular reason to allow this. I think this is where the problem lies.

**Comment by captnjameskirk** [comment]
Same here and driving me nuts, macOS 15.3 with ipython 8.32.0 and python 3.13.2.  The only "fix" I've found is to set `TerminalInteractiveShell.editing_mode=vi`, hit ESC to enter normal mode, and use j/k to move left/right instead of the arrow keys, then hit i to enter insert mode when I get to the position I want edit. Using the right arrow key in vi insert mode still has this bug, so using vi normal mode for movement within the line is the only way to avoid it.

**Comment by Carreau** [comment]
There has been a number of updates the autosuggestions, and completion in 9.0 (main), and I hve published a beta, I would appreciate for some of you to try, 

but as I can't reproduce, I can't debug. 

Please also check your prompt toolkit version; my hunch is that it's the autosuggestions, not the completions. It _should not_ be triggered when not at end of line.

**Comment by k4lizen** [comment]
@Carreau 
I can reproduce on main (a6a4b0a8)

**Comment by k4lizen** [comment]
It occurs to me almost every time I write `somefunction()` then `<left arrow>`, `<fill in the function arguments>`, `<right arrow>` ... and attempt to continue typing the rest of the line, which is instead filled with autocompleted stuff (which is always wrong, as in the 123 -> 132 case).

This is often enough that ipython is unusable for me, and I have been using regular python since encountering this bug

**Comment by Carreau** [comment]
```
In [1]: import prompt_toolkit

In [2]: prompt_toolkit.__version__
Out[2]: '3.0.49'`
```


Could you try #14823 ?

I think I managed to reproduce, but that requires a bit more setup than what is described here, 
for example it is likely that @oscarbenjamin you may have typed `solve(groebner(E, syms), syms, dict=True)` or something similar in the past ?

here is my attempt at reproducing, then trying a fix.

https://github.com/user-attachments/assets/e86fd812-e973-49ce-897e-72dda7fc8813

**Comment by Carreau** [comment]
> The 123 reproducer [#14137 (comment)](https://github.com/ipython/ipython/issues/14137#issuecomment-2332522563) doesn't work for you?

Well, it depends, I could make it to reproduce but I can make it do anything, 
like for me `1000002`

https://github.com/user-attachments/assets/6c734716-b1dc-4af3-b0a0-f7fd3f8a4460

**Comment by Carreau** [comment]
Yeah I consider it a bug (hence the PR to fix it that needs more test).

**Comment by Carreau** [comment]
closed by #14823

**Comment by oscarbenjamin** [comment]
Thanks!
